pkg/activator/net/ip_registry.go

```
@@ -0,0 +1,289 @@
+//
+// USAGE CONTRACT (Clear Separation by Authority):
+//
+//  1. Claim(ip, rev) - QP AUTHORITY (unconditional write)
+//     Queue-proxy is the authoritative source - always overwrites.
+//     Used for push-based pod registration from queue-proxy.
+//
+//  2. TryClaim(ip, rev) - INFORMER (conditional write)
+//     Informer is a distributed source - only claims unclaimed IPs.
+//     Returns false if IP is already owned by a different revision.
+//
+//  3. Owner(ip) - READ-ONLY query
+//     Returns current owner of the IP, if any
+//
+//  4. Release(ip, rev) - SOLE WRITER for removal
+//     Removes IP ownership if still owned by the revision
+//
+// PATTERN - QP Authority (push-based from queue-proxy):
+//
+//	registry.Claim(ip, revision)  // Always succeeds, overwrites previous owner
+//
+// PATTERN - Informer (K8s watch, distributed):
+//
+//	if registry.TryClaim(ip, revision) {
+//	    // Successfully claimed this IP for our revision
+//	} else {
+//	    // IP is already owned by another revision, skip it
+//	}
+type ipRevisionRegistry struct {
+	entries sync.Map // map[string]types.NamespacedName]
+	logger  *zap.SugaredLogger
+}
+
+func newIPRevisionRegistry(logger *zap.SugaredLogger) *ipRevisionRegistry {
+	if logger == nil {
+		logger = zap.NewNop().Sugar()
+	}
+	return &ipRevisionRegistry{
+		logger: logger,
+	}
+}
+
+// Claim associates the provided IP with the revision unconditionally.
+// This is the SOLE writer operation for QP Authority - always succeeds.
+// Establishes/overwrites IP ownership regardless of previous owner.
+// Used exclusively by queue-proxy push-based pod registration.
+//
+// Returns the displaced revision if this claim overwrote an existing owner,
+// or an empty NamespacedName if the IP was unclaimed or already ours.
+// This allows callers to notify the displaced revision of IP reuse.
+func (r *ipRevisionRegistry) Claim(ip string, rev types.NamespacedName) types.NamespacedName {
+	if r == nil || ip == "" {
```

this feels like it should return an error rather than fail silently. especially since empty nodespacename also means it's unclaimed.